### DISCUSSION TITLE
Is there a way to set the DB schema when creating a table?

### DISCUSSION THREAD

**Discussion by dkoehnolms** [oqtane/oqtane.framework#2251] [Q&A] [answered]
I was wanting my module’s tables to have a different schema so they could be identified in the database.

BaseEntityBuilder.cs

        public void Create()
        {
            _migrationBuilder.CreateTable(RewriteName(EntityTableName), BuildTable, null, AddKeys);
        }

The null above is the schema param. Could this be changed to a function?

**Comment by sbwalker** [comment]
I think what you are suggesting is adding a new property to BaseEntityBuilder called Schema which could be set in a derived EntityBuilder similar to the way that PrimaryKeys and ForeignKeys are currently set. The property would be used in the Create() method instead of null. And I am guessing that for those databases which do not support schemas, this would be ignored. Does this sound correct?

**Reply by dkoehnolms** [reply]
Yes

**Comment by sbwalker** [comment, marked as answer]
@dkoehnolms #2255 provides the ability to specify a Schema in a derived EntityBuilder

**Reply by nicpitsch** [reply]
@dkoehnolms did the response meet your request? Have you verified it using the latest Oqtane version?
Please update the answered status of the discussion.

**Reply by dkoehnolms** [reply]
Thanks for #2255! I have tested it on LocalDB and it worked for me. I added Schema to the model table attribute like 

```
[Table("TestModel", Schema = "test")]
public class TestModel : IAuditable
```

and added like the following to the EntityBuilder

```
 migrationBuilder.EnsureSchema("test");
 Schema = "test";
```

**Comment by sbwalker** [comment]
@dkoehnolms I am curious how you are dealing with Foreign Keys? The general pattern in the core framework is:

```
        private const string _entityTableName = "Alias";
        private readonly PrimaryKey<AliasEntityBuilder> _primaryKey = new("PK_Alias", x => x.AliasId);
        private readonly ForeignKey<AliasEntityBuilder> _tenantForeignKey = new("FK_Alias_Tenant", x => x.TenantId, "Tenant", "TenantId", ReferentialAction.Cascade);

        public AliasEntityBuilder(MigrationBuilder migrationBuilder, IDatabase database) : base(migrationBuilder, database)
        {
            EntityTableName = _entityTableName;
            PrimaryKey = _primaryKey;
            ForeignKeys.Add(_tenantForeignKey);
        }
```

However the hardcoded table names in the Foreign Key declaration would cause issues if you are using Schemas? I am curious how you are handling this?

**Reply by dkoehnolms** [reply]
Sorry, I have not needed to use Foreign Keys yet with Schemas. A Foreign Key incurs a coupling at the DB level are we willing to have the same coupling at the EntityBuilder level? Looked at the code some but did not see a clean way for this to be implemented. If you are looking at making Schemas part of the main configuration, does it fit in IDatabase?

**Reply by sbwalker** [reply]
@dkoehnolms I did some initial research into this and it appears to be a non-trivial effort - and since nobody is asking for it, I would have to prioritize it much lower than other enhancements for now.
